### ISSUE TITLE
Sik-Ho Tang | Review -- Gaussian Error Linear Units (GELUs).

### ISSUE THREAD

**Issue by NorbertZheng** [NorbertZheng/read-papers#107]
Sik-Ho Tang. [Review — Gaussian Error Linear Units (GELUs)](https://sh-tsang.medium.com/review-gaussian-error-linear-units-gelus-d4d7347d1e11).

**Comment by NorbertZheng** [comment]
## Overview
Gaussian Error Linear Units (GELUs). GELU, by University of California, and Toyota Technological Institute at Chicago. 2016 arXiv, Over 600 Citations.

Activation Unit, Image Classification, POS Tagging, Phone Recognition.

 - The GELU nonlinearity weights inputs by their value, rather than gates inputs by their sign as in ReLUs.
 - Performance improvements are obtained across all considered computer vision, natural language processing, and speech tasks.

**Comment by NorbertZheng** [comment]
## Gaussian Error Linear Unit (GELU)
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/26257859-f710-4a1c-9c1a-2d0b4efd7dfd)
GELU (μ=0, σ=0) vs ReLU vs ELU.

 - ReLU deterministically multiplying the input by zero or one and [Dropout](https://sh-tsang.medium.com/paper-dropout-a-simple-way-to-prevent-neural-networks-from-overfitting-image-classification-a74b369b4b8e) stochastically multiplying by zero.

Specifically, the neuron input x can be multiplied by $m\sim Bernoulli(\Phi(x))$, where $\Phi(x) = P(X\leq x)$; $X\sim N(0, 1)$ is the cumulative distribution function of the standard normal distribution.

 - **This distribution is chosen since neuron inputs tend to follow a normal distribution, especially with Batch Normalization.**

Since the cumulative distribution function of a Gaussian is often computed with the error function, the Gaussian Error Linear Unit (GELU) is defined as:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/92e87d05-9a25-4022-999b-ccfeada73da6)

The above equation is approximated as:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/c60fd75d-d707-4a8d-8d14-4a8e7e780a5c)

or:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/0c1cdcf8-1492-423a-8a4a-1d7f383b8da1)

if greater feedforward speed is worth the cost of exactness.

Different $N(\mu, \sigma)$ can be used as CDF, but in this paper, $N(0,1)$ is used.

**Comment by NorbertZheng** [comment]
## Experimental Results
### MNIST Classification
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/db3e9cc3-027f-49d2-9764-70af55b58736)
MNIST Classification Results. Left are the loss curves without [Dropout](https://sh-tsang.medium.com/paper-dropout-a-simple-way-to-prevent-neural-networks-from-overfitting-image-classification-a74b369b4b8e), and right are curves with a [Dropout](https://sh-tsang.medium.com/paper-dropout-a-simple-way-to-prevent-neural-networks-from-overfitting-image-classification-a74b369b4b8e) rate of 0.5.

 - Fully connected networks with 8-layer, 128 neuron wide neural network are trained for 50 epochs with a batch size of 128.

GELU tends to have the lowest median training log loss with and without [Dropout](https://sh-tsang.medium.com/paper-dropout-a-simple-way-to-prevent-neural-networks-from-overfitting-image-classification-a74b369b4b8e).

**Comment by NorbertZheng** [comment]
### MNIST Autoencoder
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/e1f218d0-f265-43d6-87c3-d75b713b6557)
MNIST Autoencoding Results.

 - A deep autoencoder is trained on MNIST using self-supervised setting.
 - The network is with layers of width 1000, 500, 250, 30, 250, 500, 1000, in order.

GELU accommodates different learning rates and significantly outperforms the other nonlinearities.

**Comment by NorbertZheng** [comment]
### TIMIT Frame Classification
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/b6ee15fb-6951-4cf1-9de4-f1396cd9fecf)
 - Phone recognition with the TIMIT dataset which has recordings of 680 speakers in a noiseless environment.
 - The system is a 5-layer, 2048-neuron wide classifier with 39 output phone labels and a [Dropout](https://sh-tsang.medium.com/paper-dropout-a-simple-way-to-prevent-neural-networks-from-overfitting-image-classification-a74b369b4b8e) rate of 0.5.

After five runs per setting, median test error chosen at the lowest validation error is 29.3% for the GELU, 29.5% for the ReLU, and 29.6% for the ELU.

**Comment by NorbertZheng** [comment]
###  CIFAR-10/100 Classification
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/1182e401-fddb-47e9-8fac-e05a78694e11)
CIFAR-10 Results.

![image](https://github.com/NorbertZheng/read-papers/assets/37235023/9cb3e025-a6e5-42f3-8ee3-876cf9a50af3)
CIFAR-100 Results.

 - A shallower convolutional neural network of 9-layer is trained to test CIFAR-10. Each curve is a median of three runs.

On CIFAR-10, ultimately, the GELU obtains a median error rate of 7.89%, the ReLU obtains 8.16%, and the ELU obtains 8.41%.

 - On CIFAR-100, WRN with 40 layers and a widening factor of 4 is trained.
 - Over three runs we obtain the median convergence curves are shown above.

On CIFAR-100, the GELU achieves a median error of 20.74%, the ReLU obtains 21.77%, and the ELU obtains 22.98%.

**Comment by NorbertZheng** [comment]
## Reference
 - [2016 arXiv] [GELU] [Gaussian Error Linear Units (GELUs)](https://arxiv.org/abs/1606.08415).
